sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -210,6 +212,223 @@ public Row getCurrentRow() throws NoSuchElementException {
+                      Row row = FlatRowConverter.convert(flatRow);
+                      currentByteSize += row.getSerializedSize();
+                      rows.add(row);
+
+                      if (currentByteSize > maxSegmentByteSize) {
+                        byteLimitReached = true;
+                        atomicScanHandler.get().cancel();
+                        return;
+                      }
+                    }
+
+                    @Override
+                    public void onError(Throwable e) {
+                      f.setException(e);
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                      ReadRowsRequest nextNextRequest = null;
+
+                      // When requested rows < limit, the current request will be the last
+                      if (byteLimitReached || rows.size() == nextRequest.getRowsLimit()) {
+                        nextNextRequest =
+                            truncateRequest(nextRequest, rows.get(rows.size() - 1).getKey());
+                      }
+                      f.set(new UpstreamResults(rows, nextNextRequest));
+                    }
+                  });
+      atomicScanHandler.set(handler);
+      return f;
+    }
+
+    private void waitReadRowsFuture() throws IOException {
+      try {
+        UpstreamResults r = future.get();
+        buffer.addAll(r.rows);
+        nextRequest = r.nextRequest;
+        future = null;
+        serviceCallMetric.call("ok");
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException(e);
+      } catch (ExecutionException e) {
+        Throwable cause = e.getCause();
+        if (cause instanceof StatusRuntimeException) {
+          serviceCallMetric.call(((StatusRuntimeException) cause).getStatus().getCode().value());
+        }
+        throw new IOException(cause);
+      }
+    }
+
+    private ReadRowsRequest truncateRequest(ReadRowsRequest request, ByteString lastKey) {
+      RowSet.Builder segment = RowSet.newBuilder();
+
+      for (RowRange rowRange : request.getRows().getRowRangesList()) {
+        int startCmp = StartPoint.extract(rowRange).compareTo(new StartPoint(lastKey, true));
+        int endCmp = EndPoint.extract(rowRange).compareTo(new EndPoint(lastKey, true));
+
+        if (startCmp > 0) {
+          // If the startKey is passed the split point than add the whole range
+          segment.addRowRanges(rowRange);
+        } else if (endCmp > 0) {
+          // Row is split, remove all read rowKeys and split RowSet at last buffered Row
+          RowRange subRange = rowRange.toBuilder().setStartKeyOpen(lastKey).build();
+          segment.addRowRanges(subRange);
+        }
+      }
+      if (segment.getRowRangesCount() == 0) {
+        return null;
+      }
+
+      ReadRowsRequest.Builder requestBuilder = request.toBuilder();
+      requestBuilder.clearRows();
```

Comment:
Yes, since we are pulling the builder from the original request, it contains the filter on Line 286. I believe that the only thing that needs to change is the RowSet under the request.